### ISSUE TITLE
Scalability of retrieving ServiceClasses

### ISSUE THREAD

**Issue by jwforres** [kubernetes-incubator/service-catalog#454]
The ServiceClass resource contains all of the metadata for its Plans, and with both of the current proposals for Parameters it will also contain the whole schema for the parameters for bind and provision.  This means each ServiceClass is potentially large, and Listing a significant number of ServiceClasses is going to create performance concerns for clients.

**Comment by pmorie** [comment]
Something we should discuss when it's time to make a `v1beta1` API

**Comment by arschles** [comment]
Moving into the beta cycle according to what @pmorie said above

**Comment by MHBauer** [comment]
Sounds like we need to make a decision whether or not to change and create a resource.

Not sure which is more kube-ish, but it looks like we would end up with a slice of refs to plans, or a ref to the existing slice of plans as a separate object.

**Comment by duglin** [comment]
I don't see a reason that we shouldn't split Plan out into its own resource.

**Comment by jwforres** [comment]
The only case that I can think of is that it makes it hard for a UI to let
a user filter the service class objects based on whether they have a free
plan available. You'd have to still request all the plans for all the
service classes, which doesn't fix the scalability concern.

**Comment by MHBauer** [comment]
@jwforres What is the common case? Is it look at the classes with no need for plans, or needing to look at the plans of a class?

If we have to look at the plans, we have to look at the plans. As long as we it isn't done that often it should be fine.

**Comment by jwforres** [comment]
I can only speak to how we are designing our particular user experience.  At least for us, most of the time you do not need anything in the plans until you pick a particular service class to look at, at which point yeah we could pull the plans for that service class.

Like I said before, being able to filter on free vs paid is the only thing I can think of that requires the plans. The problem is if you are designing a faceted filter (think amazon's filters that show up after you search) then you want to know how many results selecting a particular filter would reduce the result set down to like 

- [ ] Free (5)
- [ ] Paid (10)

To know that, we'd have to know all the plans upfront, unless you do something special to normalize the data onto the serviceclass object with something like a freePlansCount. That is probably not a road you want to go down though.

**Comment by duglin** [comment]
One option would be to have logic in our controller to calculate the UX friendly bits of data (e.g. "hasFree", "numFree", "numPaid", or even just a small subset of the plans data (free/description)) and stick those in the ServiceClass object. Then this info is easily available w/o having to pull down all of the plans.  This gives us the freedom to move the plans to a new set of resources if we think that's the best approach.

IMO getting a high-level list of services (names, description and #/types of plans) should be very quick to provide a nice UX. Once I find a service I'm interested in I can then do a deeper dive into that one.

CF's "cf marketplace" command returns 3 bits of info for each service: name, list of plan names, a short description of the service.  Each plan name has a "*" to indicate if its free or not.

Based on this what do people think about:
- each serviceClass has a slice of Plan (nested slice, not a slice of refs)
- each serviceClass.Plan has:
  - name
  - description
  - free (bool)
  - ref to the full ServicePlan resource
- put the full Plan metadata into a new (non-nested) ServicePlan resource

I'll put this on the agenda for today's design session call.

**Comment by duglin** [comment]
Per today's design call there was general agreement to split Plan from ServiceClass so each would becomes a stand-alone resource.  We also agreed to let this idea simmer for a few days so people can think it over and if no issue arise from it we will then look for a PR to be submitted to do the split so we can then use that as a starting point for our discussions around the exact layout of the two resources.  The open question include: how to manage the reference(s) between then and which properties (if any) from the Plan should be duplicated in the ServiceClass.

**Comment by duglin** [comment]
On yesterday's design call we agreed with the outcome from the previous call: split them and open a PR.

**Comment by pmorie** [comment]
I think this can be done any time.  @MHBauer are you still interested in doing this?

**Comment by MHBauer** [comment]
Yes, sorry, was looking at code for a week or two but never PR'd it as it was difficult. I was probably doing the wrong thing at the time. 

Do we need to finalize the design before moving forward anymore?

I don't think renaming blocks.

**Comment by duglin** [comment]
Copying my comment from the PR:

I think talking about this during one of our design sessions would be good.

It might also be good to try to split the discussion into smaller pieces, I think there are at least two independent topics:
1 - which properties are in Service vs Plan and which (if any) are duplicated?
2 - how to manage the references? S->P or P->S

If you could come up with a concrete proposal for each, to help focus the discussion, then we can discuss them on an upcoming call.

**Comment by MHBauer** [comment]
# Problem 
If we get a ServiceClass, we get all of the Plans, which contain a lot of data.

# Solution
 1. Split off Plans into own toplevel object. 
 1. ServiceClasses contain a slice of (LocalObjectRef / String Names). (ServiceClass -> Plan)
 1. Plans contain all data as before.

# Problems
 - Classes are not namespaced, so plans are not namespaced. Much increased possibility of collision.
 - Must get all plans individually to get plan details. (The
   alternative of all plans by asking for all plans in list would get
   all plans from all brokers.)

# Later Options
 - Create a Plan stub object consisting of the free data and pointer to Plan.
 - Have the ServiceClass have a toplevel property of the number of
   free and non-free plans. Controller reconciles this to fill it in.

**Comment by MHBauer** [comment]
Proposal updated.

**Comment by duglin** [comment]
why is:
```
ServiceClass []v1.LocalObjectReference `json:"serviceClass"`
```
an array?  Each plan just belongs to one class, right?

**Comment by MHBauer** [comment]
Yes, doug, correct.

**Comment by vaikas-google** [comment]
The remaining issue here, using field selectors namely is being addressed in #1346 so closing this one out. @pmorie @MHBauer please reopen if you feel differently.
